### PR TITLE
Only remove quotes from around the property value

### LINKED ISSUE #15: Transformer produces invalid code when string has invalid property name characters
Due to removing quotes from the strings here:

https://github.com/kimamula/ts-transformer-enumerate/blob/ea89429d4ba26fb78c302b9691c6f0f28c968e5d/transformer.ts#L24

If a string contains characters that are not allowed in a property name (such as a slash), the transformer generates invalid code.

For example: `'/users'` becomes `{ /user: '/user' }`, which is invalid.

### PR DESCRIPTION
**Pull request by G-Rath**
Fixes #15

### CODE REVIEW

**Comment by G-Rath** [comment]
I'm not sure the best/safest way to make the test pass currently.

I've tested it out, and this is the resulting output:

```
const enumerated = enumerate<'one\\two' | 'one'>();
console.log(enumerated, enumerated['one\\two'], enumerated.one);
```

Gives

`{ 'one\\two': 'one\\\\two', one: 'one' } 'one\\\\two' 'one'`.

I think the whole issue relates to the fact that Typescript runs in JS: so it gets the string, which gets unescaped, so it escapes it, but b/c it's already escaped, that results in doubling the escape...

Basically, I think it's like if you console log `'one\\\\two'`, you'll get `'one\\two'`. If you then copy that line and log that (as a string literal) you get `'one\two'`; TS is running JS through JS, resulting in a double up.

The only solution I can think of that's nice and easy would be to use `.replace('\\\\', '\\')`.

I think it might be worth doing that + stating in the README that using `\` is not recommended & a bad idea, as it leads to be confusing results.

**Comment by kimamula** [comment]
Looks like collecting literal expressions instead of strings in `resolveStringLiteralTypes()` should work.
i.e.,

```ts
function resolveStringLiteralTypes(node: ts.Node, typeChecker: ts.TypeChecker, literals: ts.LiteralTypeNode['literal'][]): void {
  switch (node.kind) {
    case ts.SyntaxKind.TypeReference:
      const symbol = typeChecker.getSymbolAtLocation((node as ts.TypeReferenceNode).typeName);
      symbol && symbol.declarations && symbol.declarations[0].forEachChild(node => resolveStringLiteralTypes(node, typeChecker, literals));
      break;
    case ts.SyntaxKind.UnionType:
      node.forEachChild(node => resolveStringLiteralTypes(node, typeChecker, literals));
      break;
    case ts.SyntaxKind.LiteralType:
      // DO NOT CALL getText()
      literals.push((node as ts.LiteralTypeNode).literal);
      break;
    default:
      break;
  }
}
```

Use these literals to create property assignment.

```ts
const literals: ts.LiteralTypeNode['literal'][] = [];
node.typeArguments && resolveStringLiteralTypes(node.typeArguments[0], typeChecker, literals);

return ts.createObjectLiteral(literals.map(literal =>
  ts.createPropertyAssignment(literal.getText(), literal)
));
```

**Comment by G-Rath** [comment]
Awesome! I'll have a play around with this later tonight, and see if I can get those testings passing :D

**Comment by G-Rath** [comment]
@kimamula that code did it - thanks!

Out of interest, where did you get that from? Did you come up with it, find it rooting around somewhere, or another place entirely?

**Comment by kimamula** [comment]
LGTM. Thanks!

> Out of interest, where did you get that from? Did you come up with it, find it rooting around somewhere, or another place entirely?

Well, I found it by exploring the type declarations of the TypeScript compiler API.
I thought we should stop stringifying string literals beforehand to prevent TypeScript from double escaping them.
Then I found that `ts.LiteralTypeNode` has a `literal` property which can be directly passed as the second argument of `ts.createPropertyAssignment()` and confirmed it actually works as expected.

**Comment by kimamula** [comment]
@G-Rath Published v0.3.2.
